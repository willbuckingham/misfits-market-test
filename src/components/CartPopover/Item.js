import React from 'react'
import PropTypes from 'prop-types'
import {
  Typography,
} from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'

import { formatCurrency } from './../../utils'

const useStyles = makeStyles(({ spacing }) => ({
  root: {
    display: 'flex',
  },
  textarea: {
    padding: spacing(1),
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    flexGrow: 1,
  },
  price: {
    display: 'flex',
    alignItems: 'center',
  },
}))

const Item = props => {
  const classes = useStyles()

  const { product, price, quantity } = props

  return (
    <>
      <div className={classes.root}>
        <img src='/fruit.jpg' alt={product} title={product} height={80} />
        <div className={classes.textarea}>
          <Typography>{product}</Typography>
          <Typography color='textSecondary'>{formatCurrency(price)} x {quantity}</Typography>
        </div>
        <div className={classes.price}>
          <Typography variant='h6'>{formatCurrency(price * quantity)}</Typography>
        </div>
      </div>
      <hr/>
    </>
  )
}

Item.propTypes = {
  product: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
  quantity: PropTypes.number.isRequired,
}

export default Item
