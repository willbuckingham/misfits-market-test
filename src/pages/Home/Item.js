import React, { useState } from 'react'
import PropTypes from 'prop-types'
import {
  // IconButton,
  Typography,
  // Button,
  Card,
  CardMedia,
  CardContent,
  CardActions,
  Fab,
  TextField,
} from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import { 
  AddShoppingCart,
} from '@material-ui/icons'

const useStyles = makeStyles(({ spacing }) => ({
  card: {
    height: '100%',
    display: 'flex',
    flexDirection: 'column',
  },
  cardContent: {
    flexGrow: 2,
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'space-between',
  },
  product: {
    lineHeight: 1.2,
  },
  priceContainer: {
    display: 'flex',
    alignItems: 'start',
    justifyContent: 'center',
    marginTop: spacing(1),
  },
  price: {
    fontWeight: 'bold',
  },
  msrp: {
    textDecoration: 'line-through',
  },
  cardActions: {
    display: 'flex',
    justifyContent: 'space-between',
    height: spacing(6),
    padding: spacing(2),
  },
}))

export const formatCurrency = v =>
  '$' + v.toFixed(2)

export const calcSavings = (msrp, price) => {
  const savingsExact = price < msrp ? (1 - price / msrp) * 100 : 0
  return Math.round(savingsExact / 5) * 5
}

const Item = props => {
  const classes = useStyles()

  const [ quantity, setQuantity ] = useState(1)

  const { id, product, price, msrp, soldOut } = props
  const savings = calcSavings(msrp, price)

  const onClickAdd = () => {
    props.onAdd({
      id,
      quantity,
    })
    setQuantity(1)
  }

  return (
    <Card className={classes.card}>

      <CardMedia
        component='img'
        image='/logo512.png'
        title={product}
        alt={product}
        height={200}
      />
      <CardContent className={classes.cardContent}>
        <Typography
          variant='h6'
          className={classes.product}
        >
          {product}
        </Typography>
        <div className={classes.priceContainer}>
          <Typography display='inline' variant='body1' color='primary' className={classes.price}>{formatCurrency(price)}</Typography>
          &nbsp;
          <Typography display='inline' variant='caption' color='textSecondary' className={classes.msrp} >{formatCurrency(msrp)}</Typography>
          &nbsp;
          { savings && <Typography display='inline' variant='caption' color='textSecondary'>{`${savings}% off MSRP`}</Typography> }
        </div>
      </CardContent>
      <CardActions className={classes.cardActions}>
        { soldOut ?
          <Typography color='error'>Sold Out!</Typography> :
          <>
            <TextField 
              value={quantity}
              onChange={ evt => setQuantity(evt.target.value) }
              label='Quantity'
              type='number'
              inputProps={{ min: 1, max: 99 }}
            />
            <Fab color='secondary' size='small' onClick={onClickAdd}>
              <AddShoppingCart/>
            </Fab>
          </>
        }
      </CardActions>
    </Card>
  )
}

Item.propTypes = {
  id: PropTypes.number.isRequired,
  product: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
  msrp: PropTypes.number.isRequired,
  soldOut: PropTypes.bool.isRequired,
}

export default Item
