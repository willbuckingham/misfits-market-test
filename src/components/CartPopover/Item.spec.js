import React from 'react'
import { shallow } from 'enzyme'

import Item from './Item'

describe('Item', () => {
  let props

  beforeAll(() => {
    props = {
      product: 'Product',
      price: 1,
      quantity: 2,
    }
  })

  it('renders', () => {
    expect(shallow(<Item { ...props } />)).toMatchSnapshot()
  })
})
