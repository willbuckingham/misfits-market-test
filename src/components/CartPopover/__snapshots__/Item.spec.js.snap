// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item renders 1`] = `
<Fragment>
  <div
    className="makeStyles-root-1"
  >
    <img
      alt="Product"
      height={80}
      src="/fruit.jpg"
      title="Product"
    />
    <div
      className="makeStyles-textarea-2"
    >
      <WithStyles(ForwardRef(Typography))>
        Product
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        color="textSecondary"
      >
        $1.00
         x 
        2
      </WithStyles(ForwardRef(Typography))>
    </div>
    <div
      className="makeStyles-price-3"
    >
      <WithStyles(ForwardRef(Typography))
        variant="h6"
      >
        $2.00
      </WithStyles(ForwardRef(Typography))>
    </div>
  </div>
  <hr />
</Fragment>
`;
